### ISSUE TITLE
2022_05_12 row counts are lower than expected

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/data-pipeline#72]
There were 8M mobile pages in the 2022_05_01 crawl. In 05_12 we enabled secondary pages, so I'd expect to see approximately 2x as many pages.

The summary table `httparchive.experimental_summary_pages.2022_05_12_mobile` only contains 5.5M pages (rows) and that's including secondary pages. The actual number of home pages is 2.8M.

The non-summary table `httparchive.pages.2022_05_12_mobile` similarly shows fewer pages than expected at 6.3M.

We could evaluate the number of HARs in `gs://httparchive/crawls/android-May_12_2022` to be sure, but it seems like we're missing a lot of data. We should investigate whether this is something that could happen again in the June crawl.

```sql
SELECT
  JSON_VALUE(metadata, '$.crawl_depth') AS crawl_depth,
  COUNT(0) AS num_rows,
  COUNT(DISTINCT url) AS distinct_pages,
FROM
  `httparchive.experimental_summary_pages.2022_05_12_mobile`
GROUP BY
  crawl_depth
```

crawl_depth | num_rows | distinct_pages
-- | -- | --
1 | 2689382 | 2594890
0 | 2776046 | 2776046

# 2022_05_12

- [ ] summary_pages mobile
- [ ] summary_requests mobile
- [x] summary_pages desktop
- [ ] summary_requests desktop
- [ ] lighthouse mobile
- [x] lighthouse desktop
- [ ] pages mobile
- [x] pages desktop
- [ ] requests mobile
- [x] requests desktop
- ~response_bodies mobile~ _(wontfix)_
- ~response_bodies desktop~ _(wontfix)_

**Comment by pmeenan** [comment]
```bash
$ gsutil du gs://httparchive/crawls/android-May_12_2022 | wc -l
 15499212
 ```
 
 I'm seeing ~15M HARs in the mobile dataset in GCS.  I don't think the missing data is from them not being tested.

**Comment by rviscomi** [comment]
That's very good news, sounds like there's no data loss and it's more of a pipeline issue. Strange that both summary and non-summary are affected, I'll investigate.

**Comment by rviscomi** [comment]
Reran the mobile + desktop summary pipeline. Mobile improved by about 1.3M pages to 8.3M total rows:

crawl_depth | num_rows | distinct_pages
-- | -- | --
1 | 4065986 | 3863425
0 | 4195732 | 4195732

Still not where I would expect it to be given the [15M HARs](https://github.com/HTTPArchive/data-pipeline/issues/72#issuecomment-1138757941) in GCS.

However, desktop went from 2.5M rows to 0.6M rows ❗📉❗ 

crawl_depth | num_rows | distinct_pages
-- | -- | --
1 | 176804 | 175959
0 | 173468 | 173468

There did happen to be a vague [HTTP error](https://console.cloud.google.com/errors/detail/CJH69_TS0e31zQE;service=Dataflow%20Jobs;version=2022-05-21_09_42_38-11634629694798666271;time=P30D?project=httparchive) at the time that the BQ write was happening. Not sure if that's the pipeline's fault or the GCP infrastructure or what.

I could use help tracking this down if anyone has the time.

**Comment by rviscomi** [comment]
Renaming this issue to reflect that all tables seem to have lower than expected row counts.

Client | Pipeline | Table | Estimated | Actual | Coverage | Size (TB) | Projected size (TB)
-- | -- | -- | -- | -- | -- | -- | --
mobile | summary | pages | 15,499,212 | 8,261,718 | 53% | 0.01 | 0.02
mobile | non-summary | pages | 15,499,212 | 6,299,483 | 41% | 0.95 | 2.35
mobile | summary | requests | 1,500,000,000 | 758,289,296 | 51% | 0.91 | 1.81
mobile | non-summary | requests | 1,500,000,000 | 546,073,979 | 36% | 4.28 | 11.76
mobile | non-summary | response_bodies | 740,397,638 | 0 | 0% | 0.00 | 40.69
mobile | non-summary | lighthouse | 15,499,212 | 6,254,034 | 40% | 1.81 | 0.00
desktop | summary | pages | 12,000,000 | 350,272 | 3% | 0.00 | 0.01
desktop | non-summary | pages | 12,000,000 | 11,278,482 | 94% | 1.87 | 1.99
desktop | summary | requests | 1,200,000,000 | 580,944,652 | 48% | 0.70 | 1.44
desktop | non-summary | requests | 1,200,000,000 | 1,086,192,455 | 91% | 8.56 | 9.46
desktop | non-summary | response_bodies | 582,345,157 | 0 | 0% | 0.00 | 32.53
desktop | non-summary | lighthouse | 12,000,000 | 11,169,130 | 93% | 3.06 | 3.29

(waiting on `gsutil` to give me an accurate desktop HAR count)

**Comment by rviscomi** [comment]
Updated the tables above and added a checklist to the top comment to track progress. Non-summary desktop tables are looking great.

I'm currently [rerunning](https://console.cloud.google.com/dataflow/jobs/us-west1/2022-05-29_07_06_15-9867760484156857728;step=;bottomTab=WORKER_LOGS;logsSeverity=INFO?project=httparchive&pageState=(%22dfTime%22:(%22l%22:%22dfJobMaxTime%22))) the non-summary mobile pipeline to test the fix in https://github.com/HTTPArchive/bigquery/pull/169. I expect to see results in about 2 hours, pass or fail.

**Comment by rviscomi** [comment]
The pipeline succeeded but the number of rows written improved, but is still not what I'd expect:

Pipeline | Table | desktop | mobile
-- | -- | -- | --
non-summary | lighthouse | 99% | 51% ❌
non-summary | pages | 100% | 51% ❌
non-summary | requests | 91% | 89% ✅ 
non-summary | response_bodies | 0% | 0% 🆗 
summary | pages | 100% | 53%
summary | requests | 48% | 51%

The non-summary mobile requests table is actually looking good. This and response_bodies are the two tables affected by the partitioning, but there are no response bodies in the May 12 crawl to write to BQ. So this is actually a very good sign that if we apply the partitioning to all tables, their full contents should be written. I'll run a new job now to test that.

**Comment by rviscomi** [comment]
Just merged https://github.com/HTTPArchive/bigquery/pull/169 which fixed the non-summary mobile tables. I've updated the tables in https://github.com/HTTPArchive/data-pipeline/issues/72#issuecomment-1139889754 to reflect the new status.

@giancarloaf also reran the summary pipeline for desktop and the requests table is now looking good as well.

Only remaining work to resolve this issue is to rerun the summary pipeline on the mobile results.

**Comment by tunetheweb** [comment]
So we're good for 1st June crawl you think?

Out of curiosity why is non-summary requests on mobile 83% now (down from 89% above) and why is that OK? Or is that a typo?

> Only remaining work to resolve this issue is to rerun the summary pipeline on the mobile results.

Just to be clear - are you doing this now? Or asking @giancarloaf to do this?

**Comment by rviscomi** [comment]
> Out of curiosity why is non-summary requests on mobile 83% now (down from 89% above) and why is that OK? Or is that a typo?

I roughly estimated that the number of requests should be about 100x the number of pages: 15M pages * 100 requests/page = 1.5B requests. The actual number 1.2B is close enough.

> Just to be clear - are you doing this now? Or asking @giancarloaf to do this?

Hoping @giancarloaf will do this tonight.

**Comment by rviscomi** [comment]
Oh, it's also lower than 89% because I deduped the table. My last Dataflow test failed but the BQ load jobs were still intact, so rather than rerunning a 4 hour Dataflow job I manually reran each BQ load job and appended the data to the tables. That was a bit haphazard and resulted in about 2x too many rows, so deduping was needed. Some requests are validly issued multiple times per page, and those dupes were dropped.

I fixed the bug that caused my test to fail so the pipeline should be working properly now, but the deduping may have resulted in fewer rows than if it had succeeded on its own. If needed, we can rerun the non-summary pipeline to get back those valid duplicates, but I think it's ok to leave it as-is.

**Comment by giancarloaf** [comment]
> Only remaining work to resolve this issue is to rerun the summary pipeline on the mobile results.

Running the job now, the `2022_05_12_mobile` tables generated under the `experimental_summary_*_2` datasets
https://console.cloud.google.com/dataflow/jobs/us-west1/2022-05-30_19_10_27-11296255570627311287?project=httparchive
